### PR TITLE
Extend the Woo GTM tracking in Calypso to also fire in the woo express stepper flow

### PR DESCRIPTION
**Pull request by dkd903**
As we don't get enough data in the Woo GTM container for trial upgrades (users going from a free woo express trial to a paid plan) yet, we should add the Woo GTM container and track users earlier in the flow. E.g. in the stepper flow here: https://wordpress.com/setup/wooexpress/

Related to # 1749-gh-Automattic/martech

## Proposed Changes

* Introduce a new method `recordGTMDatalayerEvent` and load the Woo GTM container in the same.
* Call the method in the stepper flow step changes routine

## Testing Instructions

* Install the GADebugger Chrome Extension and enable
* Switch to the branch on this PR and start Calypso
* Visit /setup/wooexpress/ and cycle through the various steps
* Open Chrome Console and notice that you can see Tag manager events for the various steps

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/lib/analytics/ad-tracking/woo/record-gtm-datalayer-event.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	if ( stepName ) {
+		loadGTMContainer( TRACKING_IDS.wooGoogleTagManagerId )
+			.then( () => initGTMContainer() )
+			.then( () => {
+				debug(
+					`recordGTMDatalayerEvent: Initialized GTM container ${ TRACKING_IDS.wooGoogleTagManagerId }`
+				);
+
+				// We ensure that we can track with GTM
+				if ( ! mayWeTrackByTracker( 'googleTagManager' ) ) {
```
Comment (gmovr) [inline client/lib/analytics/ad-tracking/woo/record-gtm-datalayer-event.ts]: Can we check consent higher up in this function, i.e. before loading and initializing GTM? I think it should work.

File: client/lib/analytics/ad-tracking/woo/record-gtm-datalayer-event.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import { debug, TRACKING_IDS } from '../constants';
+import { initGTMContainer, loadGTMContainer } from '../gtm-container';
+
+/**
+ * Sends a step event to Google Tag Manager for the woo express stepper flow.
+ *
+ * @param {string} stepName
+ * @returns {void}
+ */
+export default async function recordGTMDatalayerEvent( stepName = '' ) {
```
Comment (gmovr) [inline client/lib/analytics/ad-tracking/woo/record-gtm-datalayer-event.ts]:
```ts
export default function recordGTMDatalayerEvent( stepName: string ): void {
```

but if you want the function to be `async` you can't return `void`, and also you need to `await` for it to resolve/reject where you use the function (ex in the Stepper flow). 

e.g.

```ts
export default async function recordGTMDatalayerEvent( stepName: string ): Promise< void > { //...


// and where you use it
await recordGTMDatalayerEvent( stepName);
```

I don't think we need it to be async in this case.

File: client/lib/analytics/ad-tracking/woo/record-gtm-datalayer-event.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	if ( stepName ) {
```
Comment (gmovr) [inline client/lib/analytics/ad-tracking/woo/record-gtm-datalayer-event.ts]: as long as we set stepName as required in the type definitions, then there's no need to check it here IMO

File: client/landing/stepper/declarative-flow/trial-wooexpress-flow.ts [hunk 4]
```diff
@@ -110,6 +111,7 @@ const wooexpress: Flow = {
 
 		function submit( providedDependencies: ProvidedDependencies = {}, ...params: string[] ) {
 			recordSubmitStep( providedDependencies, intent, flowName, currentStep );
+			recordGTMDatalayerEvent( currentStep );
```
Comment (gmovr) [inline client/landing/stepper/declarative-flow/trial-wooexpress-flow.ts]: I'm not too familiar with Stepper. But I see that there's a `useSideEffect` hook (optional) that we can use, ref the stepper readme file. It says it's for tracking functions and other side-effects.

@Automattic/martech-decepticons who's been working with Stepper. Should we use this, or can we keep the tracking logic in the submit function?

**Review by gmovr — CHANGES_REQUESTED** [review]

**Comment by gmovr** [comment]
Thanks for working on this! I left some comments 👍  cc: @markbiek and @daledupreez  who've been working on this flow previously.

**Comment by gmovr** [comment]
Could you add some testing instructions as well?

File: client/lib/analytics/ad-tracking/woo/record-gtm-datalayer-event.ts (hunk 1 above)
Comment (dkd903) [reply client/lib/analytics/ad-tracking/woo/record-gtm-datalayer-event.ts]: Makes sense; better to return early if tracking is not permitted. Updated the PR.

File: client/lib/analytics/ad-tracking/woo/record-gtm-datalayer-event.ts (hunk 2 above)
Comment (dkd903) [reply client/lib/analytics/ad-tracking/woo/record-gtm-datalayer-event.ts]: @gmovr  - Apologies, that was indeed a typo :-) I copied the other method and was building this off of that other piece of code. Just before creating the PR, I might have done an undo and the async got pushed up. Thanks for spotting it.

**Comment by jeyip** [comment]
Following the testing instructions, but when I visit `/setup/woo-express` I'm redirected to my home 🤔 Is there any other environment configuration I need to do?



_Edit:_ I think I figured it out 👍 After some digging, I'm guessing that you intended to write `/setup/wooexpress` and not `/setup/woo-express` I'm, however, not seeing any onboarding steps to walk through. I'm also not seeing anything in the GA debugger 😞 

Judging from [this initial scaffolding PR's testing instructions](https://github.com/Automattic/wp-calypso/pull/71611), having non-interactable steps appears to be expected behavior, but then I wonder what exactly we're intending to track in this PR? Would you mind elaborating a bit more @dkd903?

File: client/landing/stepper/declarative-flow/trial-wooexpress-flow.ts (hunk 4 above)
Comment (jeyip) [reply client/landing/stepper/declarative-flow/trial-wooexpress-flow.ts]: I believe `useSideEffect` is called whenever a step in the flow is rendered, whereas `submit` would be called whenever the submit callback is invoked, often _to mark the completion_ of a step cc @alshakero to verify

To answer your question, it depends on whether or not you'd like to track a step being seen or a step being completed by the end user

Ex. To elaborate further on `submit`, the trial WooExpress flow has several steps ( [link](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/trial-wooexpress-flow.ts#L24-L29) ) If we look at any of the code rendering these views, we'll see the `submit` callback tied to some logic in the file

Below is the `submit` function in the `siteCreationStep`

https://github.com/Automattic/wp-calypso/blob/4144f48528e8f5d3c787376d2a05aabff3702d2c/client/landing/stepper/declarative-flow/internals/steps-repository/site-creation-step/index.tsx#L56

It's probably invoked whenever the creator of the step has decided that the `siteCreationStep` has finished

**Comment by dkd903** [comment]
> 

Thank you @jeyip You can find some additional context on this thread: p1684695879052389-slack-C1A7F9WG5

**Comment by jeyip** [comment]
In the slack thread I see that @kevinbates says "And just to clarify, we don't need an event when the free trial flow is started, only when it's completed. :slightly_smiling_face:"

In that case, I would actually move the event tracking to here above the `exitFlow` method

https://github.com/Automattic/wp-calypso/blob/8252bfb81eeeca07da01374be668d94a07355b8e/client/landing/stepper/declarative-flow/trial-wooexpress-flow.ts#L135-L137

**Comment by dkd903** [comment]
@jeyip  - I pushed a new commit to address your comment.

**Comment by jeyip** [comment]
> I'm also not seeing anything in the GA debugger 😞

Hey @dkd903, I mentioned this earlier [here](https://github.com/Automattic/wp-calypso/pull/77841/#issuecomment-1587862400), but I'm still unable to see any events fired off in the Google Analytics Debugger.

@gmovr + @dkd903 It sounds like you were both successfully seeing events triggered though? I'm assuming there's still something left in my dev environment that needs to be set up. If either of you wouldn't mind triggering the event, recording a gif, and posting it here to confirm functionality, then consider it a +1 from my end.

**Comment by dkd903** [comment]
> > I'm also not seeing anything in the GA debugger 😞
> 
> Hey @dkd903, I mentioned this earlier [here](https://github.com/Automattic/wp-calypso/pull/77841/#issuecomment-1587862400), but I'm still unable to see any events fired off in the Google Analytics Debugger.
> 
> @gmovr + @dkd903 It sounds like you were both successfully seeing events triggered though? I'm assuming there's still something left in my dev environment that needs to be set up. If either of you wouldn't mind triggering the event, recording a gif, and posting it here to confirm functionality, then consider it a +1 from my end.

Thanks @jeyip - I am AFK today, will respond to you by tomorrow, first thing.

**Comment by gmovr** [comment]
Thanks for the review. For testing ad-tracking related code, you can enable `ad-tracking` and `cookie-banner` in the development.json config file. Another option is to add `?flags=ad-tracking,cookie-banner` to the URL, however, this is less reliable, as the flag may disappear upon redirects.

Taking this for a spin now to test the functionality, after the latest changes.

**Comment by gmovr** [comment]
I tested via the GTM debugger and the browser network inspector (e.g. looking at `window.dataLayer` for the `user interaction` data layer event. Works well. 

We may have some follow-ups if firing this on this step is insufficient, but that will be fine to add later.

@dkd903 -- I restarted the test run, but if it still fails, can you rebase it? It doesn't look related, maybe just a timeout -- but better to be sure. 

cc: @kevinbates

**Comment by gmovr** [comment]
Yeah, failure in that one build task is related to some language files in stats. Rebase should fix this 👍

**Review by gmovr — APPROVED** [review]
Works well, thanks for handling the follow-ups! Let's rebase it to make sure tests pass, and it's good to go 🚢

**Comment by dkd903** [comment]
Closing this PR due to a bad merge. Opened a new PR: https://github.com/Automattic/wp-calypso/pull/78348

**Pull request closed without merging**